src/server/dflycmd.cc

```diff
@@ -370,15 +376,169 @@ void DflyCmd::ReplicaOffset(CmdArgList args, ConnectionContext* cntx) {
+    optional<SlotId> end = ReadNumeric<SlotId>(range.at_or_null("end"));
+    if (!start.has_value() || !end.has_value()) {
+      return nullopt;
+    }
+
+    ranges.push_back({.start = start.value(), .end = end.value()});
+  }
+
+  return ranges;
+}
+
+optional<ClusterConfig::Node> ParseClusterNode(const JsonType& json) {
+  if (!json.is_object()) {
+    LOG(WARNING) << kInvalidConfigPrefix << "node config is not an object " << json;
+    return nullopt;
+  }
+
+  ClusterConfig::Node node;
+
+  {
+    auto id = json.at_or_null("id");
+    if (!id.is_string()) {
+      LOG(WARNING) << kInvalidConfigPrefix << "invalid id for node " << json;
+      return nullopt;
+    }
+    node.id = std::move(id).as_string();
+  }
+
+  {
+    auto ip = json.at_or_null("ip");
+    if (!ip.is_string()) {
+      LOG(WARNING) << kInvalidConfigPrefix << "invalid ip for node " << json;
+      return nullopt;
+    }
+    node.ip = std::move(ip).as_string();
+  }
+
+  {
+    auto port = ReadNumeric<uint16_t>(json.at_or_null("port"));
+    if (!port.has_value()) {
+      return nullopt;
+    }
+    node.port = port.value();
+  }
+
+  return node;
+}
+
+optional<ClusterConfig::ClusterShards> BuildClusterConfigFromJson(const JsonType& json) {
+  ClusterConfig::ClusterShards config;
+
+  if (!json.is_array()) {
+    LOG(WARNING) << kInvalidConfigPrefix << "not an array " << json;
+    return nullopt;
+  }
+
+  for (const auto& element : json.array_value()) {
+    ClusterConfig::ClusterShard shard;
+
+    if (!element.is_object()) {
+      LOG(WARNING) << kInvalidConfigPrefix << "shard element is not an object " << element;
+      return nullopt;
+    }
+
+    auto slots = GetClusterSlotRanges(element.at_or_null("slot_ranges"));
+    if (!slots.has_value()) {
+      return nullopt;
+    }
+    shard.slot_ranges = std::move(slots).value();
+
+    auto master = ParseClusterNode(element.at_or_null("master"));
+    if (!master.has_value()) {
+      return nullopt;
+    }
+    shard.master = std::move(master).value();
+
+    auto replicas = element.at_or_null("replicas");
+    if (!replicas.is_array()) {
+      LOG(WARNING) << kInvalidConfigPrefix << "replicas is not an array " << replicas;
+      return nullopt;
+    }
+
+    for (const auto& replica : replicas.array_value()) {
+      auto node = ParseClusterNode(replica);
+      if (!node.has_value()) {
+        return nullopt;
+      }
+      shard.replicas.push_back(std::move(node).value());
+    }
+
+    config.push_back(std::move(shard));
+  }
+
+  return config;
+}
+}  // namespace
+
+void DflyCmd::ClusterConfig(CmdArgList args, ConnectionContext* cntx) {
+  SinkReplyBuilder* rb = cntx->reply_builder();
+
+  if (args.size() != 3) {
+    return rb->SendError(kSyntaxErr);
+  }
+
+  string_view json_str = ArgS(args, 2);
+  optional<JsonType> json = JsonFromString(json_str);
+  if (!json.has_value()) {
+    LOG(WARNING) << kInvalidConfigPrefix << "Can't parse JSON " << json_str;
+    return rb->SendError(kSyntaxErr);
```

Comment:
better return more details in the error here
rb->SendError("Invalid Json", kSyntaxErrType)